lib/asyn-thread.c

```diff
@@ -566,6 +566,7 @@ static void destroy_async_data(struct Curl_async *async)
         Curl_GetAddrInfoExCancel(&td->tsd.w8.cancel_ev);
         WaitForSingleObject(td->complete_ev, INFINITE);
         CloseHandle(td->complete_ev);
+        td->complete_ev = NULL;
```

Comment:
there's another closehandle a few lines above this that you could also set td->complete_ev = NULL;

Reply:
I know, but I tried to mimic existing code logic, and not to null it after each CloseHandle.
In other words, where thread_hnd is set to NULL, complete_ev should also be set to NULL. Curl_thread_destroy sets does CloseHandle, without setting it to NULL.

Reply:
Setting it to NULL might seem like a good idea, but then code flow (where thread_hnd/complete_ev are checked) will be entirely different. Perhaps, I should add a comment explaining why it's not set to NULL there. What do you guys think?

Reply:
IMO if the variables can be evaluated after destroy_async_data they should not hold the invalid values. What logic would this affect:
~~~
diff --git a/lib/asyn-thread.c b/lib/asyn-thread.c
index 4b54b0a..aa0aa95 100644
--- a/lib/asyn-thread.c
+++ b/lib/asyn-thread.c
@@ -554,11 +554,16 @@ static void destroy_async_data(struct Curl_async *async)
 
     if(!done) {
 #ifdef _WIN32
-      if(td->complete_ev)
+      if(td->complete_ev) {
         CloseHandle(td->complete_ev);
+        td->complete_ev = NULL;
+      }
       else
 #endif
-      Curl_thread_destroy(td->thread_hnd);
+      {
+        Curl_thread_destroy(td->thread_hnd);
+        td->thread_hnd = curl_thread_t_null;
+      }
     }
     else {
 #ifdef _WIN32
~~~

Reply:
In this case the change would be OK, as long as others agree. This would be somewhat unrelated change (updating thread_hnd). Do you think this should be done as part of this PR? If yes, then I'll update

Reply:
> What logic would this affect

Well, it's not that clear. That entire function is quite mysterious. Most of the logic seemingly happens inside the block that should rarely execute. At the top of the block it goes: 

```
struct thread_data *td = async->tdata;
...
if(!done) {
  Curl_thread_destroy(td->thread_hnd);
}
else {
  Curl_thread_join(&td->thread_hnd);
  destroy_thread_sync_data(&td->tsd);
  free(async->tdata);
}
async->tdata = NULL;
...
```

it looks as if it should have been `if (done)` instead of `if (!done)`

Reply:
the only time done=1 is when starting a thread of creating the complete_ev fails. Makes no sense to try to execute that else block that curl has now.

Reply:
@Ch40zz can you please flip that done condition locally, do you still get the error with asan?

Reply:
> If you are uncomfortable with it I can submit it separately as a PR so it goes in before this and then you can rebase
I'm OK with this, I was just asking if you are ok with it